### PR TITLE
feat: batchUpsert fix#107

### PR DESCRIPTION
**Pull request by JimmyDaddy**
updateOnDuplicate
   * see https://dev.mysql.com/doc/refman/8.0/en/insert-on-duplicate.html
   * see https://www.postgresql.org/docs/9.5/sql-insert.html
   * see https://www.sqlite.org/lang_UPSERT.html
   
```SQL
## Postgres SQLite
INSERT INTO `tableName` (`name`)  VALUES ('foo'),('bar') ON CONFLICT (`name`)  DO UPDATE SET `name`=EXCLUDED.`name`;

## MySQL
INSERT INTO `tableName` (`name`) VALUES ('foo'),('bar') ON DUPLICATE KEY UPDATE `name`=VALUES(`name`);
```

```js
// MySQL
User.bulkCreate(records, { updateOnDuplicate: true });
User.bulkCreate(records, { updateOnDuplicate: ['name', 'email'] });

// Postgres SQLite
User.bulkCreate(records, { updateOnDuplicate: true });
User.bulkCreate(records, { updateOnDuplicate: ['name', 'email'] });
User.bulkCreate(records, { updateOnDuplicate: ['name', 'email'], uniqueKeys: ['email'] });
```

### CODE REVIEW

File: lib/drivers/abstract/schema.js [hunk 1]
```diff
@@ -163,10 +169,77 @@ module.exports = {
...
+            for (const field of opts.conflictKeys) {
+              const attribute = escapeId(unalias(field));
+              conflictKeys.push(attribute);
+            }
+          } else {
+            conflictKeys.push(escapeId(unalias(opts.conflictKeys)));
+          }
+        } else {
+          // conflict_target must be unique
+          // get all unique keys
```
Comment (JimmyDaddy) [inline lib/drivers/abstract/schema.js:205]: How to get all unique indexes?

File: test/models/user.js [hunk 2]
```diff
@@ -25,7 +25,10 @@ class User extends Bone {
 
 // test init
 User.init({
-  id: DataTypes.BIGINT,
+  id: {
+    primaryKey: true,
```
Comment (JimmyDaddy) [inline test/models/user.js:29]: Cant detect primary key automatically

File: lib/drivers/abstract/schema.js [hunk 3]
```diff
@@ -128,13 +128,74 @@ module.exports = {
...
+          if (att.unique) {
+            uniqueKeys.push(escapeId(att.columnName));
+            break;
+          }
+        }
+      }
+      const { primaryKey } = opts;
+      if (!uniqueKeys.length) uniqueKeys.push(escapeId(primaryKey));
+      // default use id as primary key
+      if (!uniqueKeys.length) uniqueKeys.push(escapeId('id'));
```
Comment (cyjake) [inline lib/drivers/abstract/schema.js:170]: Looks like it's possible to push `id` into `uniqueKeys` twice? Will that cause sql error?

File: lib/drivers/abstract/schema.js [hunk 4]
```diff
@@ -128,13 +128,74 @@ module.exports = {
...
+    }
+
+    if (Array.isArray(opts.updateOnDuplicate) && opts.updateOnDuplicate.length) {
+      for (const field of opts.updateOnDuplicate) {
+        const attribute = escapeId(field);
+        updateKeys.push(`${attribute}=EXCLUDED.${attribute}`);
+      }
+    } else {
+      for (const attribute of attributes) {
+        updateKeys.push(`${attribute.columnName}=EXCLUDED.${attribute.columnName}`);
```
Comment (cyjake) [inline lib/drivers/abstract/schema.js:180]: `EXCLUDED` should be commented here, since it's PostgreSQL only.

Comment (cyjake) [reply lib/drivers/abstract/schema.js:180]: Theoretically speaking, the default vendor in `lib/drivers/abstract/schema.js` should be MySQL or ANSI-SQL

File: lib/drivers/abstract/schema.js [hunk 5]
```diff
@@ -128,13 +128,74 @@ module.exports = {
...
+
+    if (opts.uniqueKeys) {
+      if (Array.isArray(opts.uniqueKeys) && opts.uniqueKeys.length) {
+        for (const field of opts.uniqueKeys) {
+          const attribute = escapeId(field);
+          uniqueKeys.push(attribute);
+        }
+      } else {
+        uniqueKeys.push(escapeId(opts.uniqueKeys));
+      }
```
Comment (cyjake) [inline lib/drivers/abstract/schema.js:153]:
```suggestion
     for (const field of [].concat(opts.uniqueKeys)) {
       uniqueKeys.push(escapeId(field));
      }
```

File: test/unit/drivers/sqlite.test.js [hunk 6]
```diff
@@ -73,7 +73,7 @@ describe('=> SQLite driver', () => {
       id: { type: BIGINT, primaryKey: true, autoIncrement: true },
       title: { type: STRING, allowNull: false },
     });
-    await driver.query(`INSERT INTO notes (id, title) VALUES (42, 'Untitled')`);
+    await driver.query('INSERT INTO notes (id, title) VALUES (42, \'Untitled\')');
```
Comment (cyjake) [inline test/unit/drivers/sqlite.test.js:76]: These changes are a bit unnecessary. Can the corresponding eslint rule be adjusted?

File: lib/drivers/mysql/schema.js [hunk 7]
```diff
@@ -77,4 +77,24 @@ module.exports = {
...
+    let attrStr;
+    const { escapeId } = this;
+    const { attributes } = opts;
+    if (Array.isArray(opts.updateOnDuplicate) && opts.updateOnDuplicate.length) {
+      attrStr = opts.updateOnDuplicate.map((field) => {
+        const attribute = escapeId(field);
+        return `${attribute}=VALUES(${attribute})`;
+      }).join(',');
+    } else {
+      attrStr = attributes.map((attribute) => `${attribute.columnName}=VALUES(${attribute.columnName})`).join(',');
```
Comment (cyjake) [inline lib/drivers/mysql/schema.js:96]: `columnName` should be escaped as well

Comment (cyjake) [inline lib/drivers/mysql/schema.js:96]:
```suggestion
      attrStr = attributes.map((attribute) => {
        const columnStr = escapeId(attribute.columnName);
        return `${columnStr}=VALUES(${columnStr})`
      }).join(',');
```

File: lib/drivers/abstract/schema.js (hunk 4 above)
Comment (cyjake) [reply lib/drivers/abstract/schema.js:180]: seems SQLite shares same syntax as PostgreSQL, agreed on current arrangement.

**Pull request merged**